### PR TITLE
proof definitions embedded in Rust docs

### PR DESCRIPTION
**Pull request by Shoeboxam**
Added documentation and proof definitions to the rust docs in the metrics, domains, measures, core, traits and traits/samplers modules. I also moved a few traits.
The transformations, measurements and combinators modules don't have proof definitions yet.

I put up a copy of the docs site up to show how the Katex rendering looks.
metrics: http://shoeboxam.github.io/opendp/metrics/index.html
domains: http://shoeboxam.github.io/opendp/domains/index.html
measures: http://shoeboxam.github.io/opendp/measures/index.html
traits: http://shoeboxam.github.io/opendp/traits/index.html
traits/samplers: http://shoeboxam.github.io/opendp/traits/samplers/index.html

To build on your own:
```shell
cd rust
cargo rustdoc --features untrusted,derive -- --html-in-header opendp_tooling/katex.html
```
Html files are located in `rust/target/doc/opendp`

### CODE REVIEW

File: rust/src/traits/samplers/geometric/mod.rs [hunk 1]
```diff
@@ -9,12 +9,19 @@ use crate::{
...
+    /// # Proof Definition
     /// Sample from the censored geometric distribution with parameter `prob`.
     /// If `trials` is None, there are no timing protections, and the support is:
+    /// ```text
     ///     [Self::MIN, Self::MAX]
+    /// ```
+    /// 
     /// If `trials` is Some, execution runs in constant time, and the support is
+    /// ```text
     ///     [Self::MIN, Self::MAX] ∩ {shift ±= {0, 1, 2, ..., `trials`}}
```
Comment (andrewvyrros) [inline rust/src/traits/samplers/geometric/mod.rs:23]: Is this ±= or just ±?

And why does trials have backticks?

File: rust/src/traits/samplers/geometric/mod.rs [hunk 2]
```diff
@@ -95,7 +102,9 @@ where
     }
 }
 
+/// Sample from the censored two-sided geometric distribution.
 pub trait SampleDiscreteLaplaceLinear<P>: SampleGeometric<P> {
+    /// # Proof Definition
```
Comment (andrewvyrros) [inline rust/src/traits/samplers/geometric/mod.rs:107]: Use a similar format as `sample_geometric` above, using text blocks? Or vice versa?

File: rust/src/traits/samplers/uniform/mod.rs [hunk 3]
```diff
@@ -12,30 +12,14 @@ use num::One;
...
-    /// # Arguments
-    ///
-    /// `min`: f64 minimum of uniform distribution (inclusive)
-    /// `max`: f64 maximum of uniform distribution (non-inclusive)
-    ///
-    /// # Return
-    /// Random draw from Unif[min, max).
+    /// # Proof Definition
+    /// Return `Err(e)` if there is insufficient system entropy, or
+    /// `Ok(sample)`, where `sample` a draw from Uniform[0,1).
```
Comment (andrewvyrros) [inline rust/src/traits/samplers/uniform/mod.rs:19]:
```suggestion
    /// `Ok(sample)`, where `sample` is a draw from Uniform[0,1).
```

File: rust/src/traits/samplers/bernoulli/mod.rs [hunk 4]
```diff
@@ -10,7 +10,11 @@ use super::{fill_bytes, sample_geometric_buffer};
 #[cfg(feature = "use-mpfr")]
 use super::SampleUniformIntBelow;
 
+/// Sample from `Bernoulli(p=0.5)`.
 pub trait SampleStandardBernoulli: Sized {
+    /// # Proof Definition
+    /// Return `Err(e)` if there is insufficient system entropy, or
+    /// `Ok(sample)`, where `sample` a draw from `Bernoulli(p=0.5)`.
```
Comment (andrewvyrros) [inline rust/src/traits/samplers/bernoulli/mod.rs:17]:
```suggestion
    /// `Ok(sample)`, where `sample` is a draw from `Bernoulli(p=0.5)`.
```

File: rust/src/traits/samplers/uniform/mod.rs [hunk 5]
```diff
@@ -48,11 +32,25 @@ pub trait SampleUniform: Sized {
...
+    ///
+    /// The idea behind the uniform sampling is to first sample a "precision band".
+    /// Each band is a range of floating point numbers with the same level of arithmetic precision
+    /// and is situated between powers of two.
+    /// A band is sampled with probability relative to the unit of least precision using the Geometric distribution.
+    /// That is, the uniform sampler will generate the band [1/2,1) with probability 1/2, [1/4,1/2) with probability 1/4,
+    /// and so on.
+    ///
+    /// Once the precision band has been selected, floating numbers numbers are generated uniformly within the band
+    /// by generating a 52-bit mantissa uniformly at random.
```
Comment (andrewvyrros) [inline rust/src/traits/samplers/uniform/mod.rs:53]: Not sure if anything can be done about this, but this block doesn't seem to appear anywhere in the rustdoc.

File: rust/src/traits/samplers/uniform/mod.rs [hunk 6]
```diff
@@ -116,14 +119,20 @@ macro_rules! impl_sample_mantissa {
 impl_sample_mantissa!(f64, 0b00001111);
 impl_sample_mantissa!(f32, 0b01111111);
 
-
+/// Sample an integer uniformly over `[Self::MIN, upper]`.
```
Comment (andrewvyrros) [inline rust/src/traits/samplers/uniform/mod.rs:122]:
```suggestion
/// Sample an integer uniformly over `[Self::MIN, Self::MAX]`.
```

File: rust/src/transformations/sum/int/monotonic/mod.rs [hunk 7]
```diff
@@ -1,12 +1,12 @@
-use num::Zero;
+
```
Comment (andrewvyrros) [inline rust/src/transformations/sum/int/monotonic/mod.rs:1]:
```suggestion
```

**Review by andrewvyrros — COMMENTED** [review]
Great enhancement of the docs! This adds a lot of helpful contexts.

Many of the comments are nitpicks, feel free to ignore.

File: rust/src/measures/mod.rs [hunk 8]
```diff
@@ -6,7 +17,19 @@ use std::{
...
+/// The greatest divergence between any randomly selected subset of the support.
+/// 
+/// # Proof Definition
+/// 
+/// ### `d`-closeness
+/// For any two vectors $u, v \in \texttt{D}$ and any $d$ of generic type $\texttt{Q}$, 
+/// we say that $u, v$ are $d$-close under the max divergence measure (abbreviated as $D_{\infty}$) whenever
+/// 
+/// ```math
+/// D_{\infty}(M(u) \| M(v)) = \max_{S \subseteq \textrm{Supp}(Y)} \Big[\ln \dfrac{\Pr[M(u) \in S]}{\Pr[M(v) \in S]} \Big] \leq d.
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/measures/mod.rs:31]: Absolutely!

File: rust/src/metrics/mod.rs [hunk 9]
```diff
@@ -1,15 +1,50 @@
...
+/// 
+/// # Proof Definition
+/// 
+/// ### `d`-closeness
+/// For any two vectors $u, v \in \texttt{D}$ and any $d$ of type [`IntDistance`], 
+/// we say that $u, v$ are $d$-close under the symmetric distance metric 
+/// (abbreviated as $d_{Sym}$) whenever 
+/// 
+/// ```math
+/// d_{Sym}(u, v) = |MultiSet(u) \Delta MultiSet(v)| \leq d
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/metrics/mod.rs:38]: Members of VectorDomain are currently defined as vectors, and the SymmetricDistance metric converts to multisets, where the symmetric difference is defined. But it does introduce baggage on the definition of the symmetric distance. This seems like a detail that should belong in the domain-metric pairing, which we don't currently model.

File: rust/src/traits/samplers/uniform/mod.rs (hunk 5 above)
Comment (Shoeboxam) [reply rust/src/traits/samplers/uniform/mod.rs:53]: I moved it up to the trait impl. Good to keep in mind in the future.

**Review by andrewvyrros — APPROVED** [review]
I think CollectionSize works better than CollectionDomain.

**Pull request merged**